onecodex/scripts/subset_reads.py

```diff
@@ -0,0 +1,387 @@
+    """
+
+    tax_id_map = {}
+
+    if parent:
+        for row in classification.results()['table']:
+            if row['parent_tax_id'] is not None and \
+               row['tax_id'] is not None:
+                tax_id_map[row['tax_id']] = row['parent_tax_id']
+    else:
+        for row in classification.results()['table']:
+            if row['parent_tax_id'] is not None and \
+               row['tax_id'] is not None:
+                try:
+                    tax_id_map[row['parent_tax_id']].add(row['tax_id'])
+                except KeyError:
+                    tax_id_map[row['parent_tax_id']] = set([row['tax_id']])
+
+    return tax_id_map
+
+
+def recurse_taxonomy_map(tax_id_map, tax_id, parent=False):
+    """
+    Takes the output dict from make_taxonomy_map and an input tax_id
+    and recurses either up or down through the tree to get /all/ children
+    (or parents) of the given tax_id.
+    """
+
+    if parent:
+        # TODO: allow filtering on tax_id and its parents, too
+        pass
+    else:
+        def _child_recurse(tax_id, visited):
+            try:
+                children = [tax_id] + list(tax_id_map[tax_id])
+            except KeyError:
+                children = [tax_id]
+
+            for child in children:
+                if child not in visited:
+                    visited.append(child)
+                    children.extend(_child_recurse(child, visited))
+
+            return children
+
+        return list(set(_child_recurse(tax_id, [])))
+
+
+def too_many_fastx_records():
+    raise ValidationError('FASTX file(s) provided have more records than the classification results')
+
+
+@click.command('subset_reads',
+               help='Subset a FASTX file based on the taxonomic results from a CLASSIFICATION_ID. '
+                    'By default, reads (or pairs or reads) matching the given taxonomic ID are '
+                    'written to the path provided, ignoring low confidence taxonomic assignments. ')
+@click.argument('classification_id')
+@click.argument('fastx', type=click.Path())
+@click.option('-t', '--tax-id', 'tax_ids', required=True, multiple=True,
+              help='Subset reads mapping to tax IDs. May be passed multiple times.')
+@click.option('-r', '--reverse', type=click.Path(),
+              help='The reverse (R2) read file, optionally.')
+@click.option('--validate/--no-validate', is_flag=True, default=True,
+              help='By default, validate FASTQ records (slow). Disable to drastically increase speed')
+@click.option('--with-children', default=False, is_flag=True,
+              help='Match child taxa of those given with -t (e.g., all strains of E. coli)')
+@click.option('--split-pairs', default=False, is_flag=True,
+              help='By default, if either read in a pair matches, both will match. Choose this '
+                   'option to consider each paired-end read separately. Resulting files may *not* '
+                   'have the same number of reads!')
```

Comment:
Let's call it `--subset-pairs-independently` and have the following help message:

```
By default, if either read in a pair matches, both will be retained in the subset file. With this option, R1 and R2 files will be evaluated independently. Note that the subset output FASTQs are *not* guaranteed to have the same number of reads!
```

Reply:
Changed as requested.